File: sample-tests/kotlin/src/main/kotlin/com/icure/docs/generator/MdxExtractor.kt
package com.icure.docs.generator

data class ExtractedBlock(
    val code: String,
    val startLine: Int,
    val testId: String? = null,
)

private val TEST_ID_REGEX = Regex("""\btest-(\w{4})\b""")

fun extractKotlinBlocks(mdxContent: String): List<ExtractedBlock> {
    val lines = mdxContent.split("\n")
    val blocks = mutableListOf<ExtractedBlock>()
    var i = 0

    while (i < lines.size) {
        val trimmed = lines[i].trim()

        if (trimmed.startsWith("```kotlin")) {
            val fenceInfo = trimmed.removePrefix("```")

            if ("no-test" in fenceInfo) {
                i++
                while (i < lines.size && !lines[i].trim().startsWith("```")) i++
                i++
                continue
            }

            val testId = TEST_ID_REGEX.find(fenceInfo)?.groupValues?.get(1)
            val startLine = i + 1
            val codeLines = mutableListOf<String>()
            i++

            while (i < lines.size && !lines[i].trim().startsWith("```")) {
                codeLines.add(lines[i])
                i++
            }
            i++

            blocks.add(ExtractedBlock(codeLines.joinToString("\n"), startLine, testId))
        } else {
            i++
        }
    }

    return blocks
}

fun isKotlinImportLine(line: String): Boolean {
    val trimmed = line.trim()
    if (trimmed.startsWith("//")) return false
    return trimmed.startsWith("import ") || trimmed.startsWith("import\t")
}

/**
 * Strip Kotlin modifiers that are invalid inside function bodies:
 * - `private`, `public`, `internal`, `protected` on val/var/fun/const
 * - `const` on local variables
 */
private val MODIFIER_STRIP_REGEX = Regex(
    """^(\s*)(?:(?:private|public|internal|protected)\s+)?(?:const\s+)?((?:val|var|fun|suspend)\s)"""
)

fun stripImportsFromBlock(code: String): String {
    val lines = code.split("\n")
    val bodyLines = lines.filterNot { isKotlinImportLine(it) }

    val trimmed = bodyLines
        .dropWhile { it.isBlank() }
        .dropLastWhile { it.isBlank() }

    // Strip modifiers invalid in local scope
    val strippedLines = trimmed.map { line ->
        MODIFIER_STRIP_REGEX.replace(line) { match ->
            match.groupValues[1] + match.groupValues[2]
        }
    }

    return strippedLines.joinToString("\n")
}

File: sample-tests/typescript/src/mdx-extractor.ts
export interface ExtractedBlock {
  code: string
  startLine: number
  /** Stable 4-letter test ID from `// @test: XXXX` on the first code line */
  testId?: string
}

export function extractTypeScriptBlocks(mdxContent: string): ExtractedBlock[] {
  const lines = mdxContent.split('\n')
  const blocks: ExtractedBlock[] = []
  let i = 0

  while (i < lines.length) {
    const line = lines[i]
    const trimmed = line.trim()

    if (trimmed.startsWith('```typescript')) {
      const fenceInfo = trimmed.slice(3)

      if (fenceInfo.includes('no-test')) {
        i++
        while (i < lines.length && !lines[i].trim().startsWith('```')) {
          i++
        }
        i++
        continue
      }

      // Extract test ID from fence info: ```typescript test-XXXX
      const testIdMatch = fenceInfo.match(/\btest-(\w{4})\b/)
      const testId = testIdMatch?.[1]

      const startLine = i + 1
      const codeLines: string[] = []
      i++

      while (i < lines.length && !lines[i].trim().startsWith('```')) {
        codeLines.push(lines[i])
        i++
      }
      i++

      blocks.push({
        code: codeLines.join('\n'),
        startLine,
        testId,
      })
    } else {
      i++
    }
  }

  return blocks
}

export function isImportLine(line: string): boolean {
  const trimmed = line.trim()
  if (trimmed.startsWith('//')) return false
  return (trimmed.startsWith('import ') || trimmed.startsWith('import\t')) && !trimmed.startsWith('import(')
}

/**
 * Remove import statements from a code block, including multi-line imports.
 * Strips leading/trailing blank lines from the result.
 */
export function stripImportsFromBlock(code: string): string {
  const lines = code.split('\n')
  const bodyLines: string[] = []
  let i = 0

  while (i < lines.length) {
    const trimmed = lines[i].trim()

    if (isImportLine(lines[i])) {
      // Check if this is a multi-line import (has '{' but no '}')
      if (trimmed.includes('{') && !trimmed.includes('}')) {
        // Skip all lines until we find the closing '}'
        i++
        while (i < lines.length && !lines[i].includes('}')) {
          i++
        }
        i++ // skip the closing '}' line
      } else {
        i++ // single-line import
      }
    } else {
      bodyLines.push(lines[i])
      i++
    }
  }

  while (bodyLines.length > 0 && bodyLines[0].trim() === '') bodyLines.shift()
  while (bodyLines.length > 0 && bodyLines[bodyLines.length - 1].trim() === '') bodyLines.pop()

  // Strip leading `export` keyword from declarations (invalid inside function bodies)
  const strippedLines = bodyLines.map(line => line.replace(/^(\s*)export\s+(async\s+function|function|const|let|class)\s/, '$1$2 '))

  return strippedLines.join('\n')
}
